In [21]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
sns.set()
pd.options.display.max_rows = 200
pd.options.display.max_columns = 100

**Taking only including 2018 to have full years**, possibly will add 2019 later.
**TODO1** - put a link to a specific file working on from drive

In [3]:
markers_raw = pd.read_csv('../../views_and_main_tables_2019_11/markers_hebrew.csv')
m_all = markers_raw[markers_raw['accident_year'] < 2019]

In [22]:
m_all.head()

id  provider_and_id  provider_code  \
0  2009000007      12009000007              1   
1  2012000013      12012000013              1   
2  2012023095      12012023095              1   
3  2012023101      12012023101              1   
4  2011000080      12011000080              1   

                   provider_code_hebrew  file_type_police  accident_type  \
0  הלשכה המרכזית לסטטיסטיקה - סוג תיק 1               NaN              1   
1  הלשכה המרכזית לסטטיסטיקה - סוג תיק 1               NaN              1   
2  הלשכה המרכזית לסטטיסטיקה - סוג תיק 1               NaN              9   
3  הלשכה המרכזית לסטטיסטיקה - סוג תיק 1               NaN              9   
4  הלשכה המרכזית לסטטיסטיקה - סוג תיק 1               NaN              1   

           accident_type_hebrew  accident_severity accident_severity_hebrew  \
0               פגיעה בהולך רגל                  3                      קלה   
1               פגיעה בהולך רגל                  3                      קלה   
2  ירידה מהכביש או עלייה למדרכה                  2                      קשה   
3  ירידה מהכביש או עלייה למדרכה                  3                      קלה   
4               פגיעה בהולך רגל                  3                      קלה   

    accident_timestamp  location_accuracy location_accuracy_hebrew  road_type  \
0  2009-02-06 00:00:00                  1              עיגון מדויק          2   
1  2012-02-01 00:00:00                  2                מרכז ישוב          2   
2  2012-12-24 11:30:00                  1              עיגון מדויק          4   
3  2012-01-20 11:30:00                  1              עיגון מדויק          4   
4  2011-06-15 00:00:00                  2                מרכז ישוב          2   

      road_type_hebrew  road_shape road_shape_hebrew  day_type  \
0     עירונית לא בצומת           9               אחר         4   
1     עירונית לא בצומת           9               אחר         4   
2  לא-עירונית לא בצומת           9               אחר         4   
3  לא-עירונית לא בצומת           9               אחר         4   
4     עירונית לא בצומת           8   כביש ישר / צומת         4   

  day_type_hebrew  police_unit police_unit_hebrew  one_lane  \
0         יום אחר           52          מרחב שפלה         9   
1         יום אחר           43         מרחב יהודה         9   
2         יום אחר           34           מרחב נגב         0   
3         יום אחר           14         מרחב עמקים         2   
4         יום אחר           34           מרחב נגב         9   

            one_lane_hebrew  multi_lane     multi_lane_hebrew  speed_limit  \
0        לא ידוע מס מסלולים           0                   NaN            0   
1        לא ידוע מס מסלולים           0                   NaN            0   
2                       NaN           2  מיפרדה עם גדר בטיחות            6   
3  דו סיטרי + קו הפרדה רצוף           0                   NaN            4   
4        לא ידוע מס מסלולים           0                   NaN            1   

  speed_limit_hebrew  road_intactness road_intactness_hebrew  road_width  \
0            לא ידוע                0                לא ידוע           0   
1            לא ידוע                0                לא ידוע           0   
2           100 קמ"ש                1            אין ליקויים           3   
3            80 קמ"ש                0                לא ידוע           3   
4         עד 50 קמ"ש                0                לא ידוע           0   

  road_width_hebrew  road_sign road_sign_hebrew  road_light road_light_hebrew  \
0           לא ידוע          5          לא ידוע           6      לילה לא ידוע   
1           לא ידוע          5          לא ידוע           6      לילה לא ידוע   
2     7 עד 10.5 מטר          4    לא נדרש תמרור           1      אור יום רגיל   
3     7 עד 10.5 מטר          3        אין ליקוי           1      אור יום רגיל   
4           לא ידוע          5          לא ידוע           6      לילה לא ידוע   

   road_control road_control_hebrew  weather weather_hebrew  road_surface  \
0           NaN                

In [4]:
m_all.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 720541 entries, 0 to 761517
Data columns (total 98 columns):
id                                           720541 non-null int64
provider_and_id                              720541 non-null int64
provider_code                                720541 non-null int64
provider_code_hebrew                         720541 non-null object
file_type_police                             55400 non-null float64
accident_type                                720541 non-null int64
accident_type_hebrew                         720541 non-null object
accident_severity                            720541 non-null int64
accident_severity_hebrew                     720541 non-null object
accident_timestamp                           720541 non-null object
location_accuracy                            720541 non-null int64
location_accuracy_hebrew                     720541 non-null object
road_type                                    720541 non-null int64
road_type_he

In [5]:
m_all.nunique()

id                                           706206
provider_and_id                              720495
provider_code                                     2
provider_code_hebrew                              2
file_type_police                                  2
accident_type                                    19
accident_type_hebrew                             21
accident_severity                                 3
accident_severity_hebrew                          3
accident_timestamp                           248993
location_accuracy                                 5
location_accuracy_hebrew                          5
road_type                                         4
road_type_hebrew                                  4
road_shape                                       13
road_shape_hebrew                                12
day_type                                          4
day_type_hebrew                                   4
police_unit                                      14
police_unit_

## Split up into main sub-frames

Understand what's similar about columns where only small part of data is missing

In [7]:
m_pol = m_all[m_all['provider_code'] == 1]  # police-investigated accidents
m_no_pol = m_all[m_all['provider_code'] == 3] # not police-investigated accidents

**Information about accidents investigated by police**

In [12]:
m_pol.info()

Information about accidents INVESTIGATED by Police
<class 'pandas.core.frame.DataFrame'>
Int64Index: 154581 entries, 0 to 760753
Data columns (total 98 columns):
id                                           154581 non-null int64
provider_and_id                              154581 non-null int64
provider_code                                154581 non-null int64
provider_code_hebrew                         154581 non-null object
file_type_police                             12557 non-null float64
accident_type                                154581 non-null int64
accident_type_hebrew                         154581 non-null object
accident_severity                            154581 non-null int64
accident_severity_hebrew                     154581 non-null object
accident_timestamp                           154581 non-null object
location_accuracy                            154581 non-null int64
location_accuracy_hebrew                     154581 non-null object
road_type                   

In [9]:
m_pol.nunique()

id                                           154577
provider_and_id                              154577
provider_code                                     1
provider_code_hebrew                              1
file_type_police                                  2
accident_type                                    19
accident_type_hebrew                             21
accident_severity                                 3
accident_severity_hebrew                          3
accident_timestamp                           115922
location_accuracy                                 5
location_accuracy_hebrew                          5
road_type                                         4
road_type_hebrew                                  4
road_shape                                       11
road_shape_hebrew                                12
day_type                                          4
day_type_hebrew                                   4
police_unit                                      14
police_unit_

**Information about accidents NOT investigated by police**

In [10]:
m_no_pol.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 565960 entries, 5 to 761517
Data columns (total 98 columns):
id                                           565960 non-null int64
provider_and_id                              565960 non-null int64
provider_code                                565960 non-null int64
provider_code_hebrew                         565960 non-null object
file_type_police                             42843 non-null float64
accident_type                                565960 non-null int64
accident_type_hebrew                         565960 non-null object
accident_severity                            565960 non-null int64
accident_severity_hebrew                     565960 non-null object
accident_timestamp                           565960 non-null object
location_accuracy                            565960 non-null int64
location_accuracy_hebrew                     565960 non-null object
road_type                                    565960 non-null int64
road_type_he

In [15]:
m_no_pol.nunique()

id                                           565918
provider_and_id                              565918
provider_code                                     1
provider_code_hebrew                              1
file_type_police                                  1
accident_type                                    19
accident_type_hebrew                             21
accident_severity                                 1
accident_severity_hebrew                          1
accident_timestamp                           220448
location_accuracy                                 5
location_accuracy_hebrew                          5
road_type                                         4
road_type_hebrew                                  4
road_shape                                       12
road_shape_hebrew                                10
day_type                                          4
day_type_hebrew                                   4
police_unit                                      13
police_unit_

## Columns where only some values are missing

Hebrew fields not matching the number field (by nulls, or by values):
- road_shape_hebrew - by number of nulls
- one_lane_hebrew - by number of nulls
- multi_lane_hebrew - by number of nulls
- speed_limit_hebrew - by number of nulls
- object_distance_hebrew - by number of nulls
- didnt_cross_hebrew - by number of nulls
- road_control_hebrew - by number of nulls
- cross_mode_hebrew - by number of nulls
- cross_location_hebrew - by number of nulls
- region_hebrew - by number of nulls
- district_hebrew - by number of nulls
- natural_area_hebrew - by number of nulls
- municipal_status_hebrew - by number of nulls
- yishuv_shape_hebrew - by number of nulls
- street1_hebrew - by number of nulls
- street2_hebrew - by number of nulls
- non_urban_intersection_hebrew - by number of nulls
- accident_hour_raw_hebrew - by number of nulls

`hebrew` fields are important and are used, they are produced from dictionaries from numeric values.  Therefore it's a problem that they don't match the numeric fields.  **TODO:** understand where missing

Comparing `markers.csv` and `markers_hebrew.csv` number of non-null values:
- Number of not-null instances seems the same in both files for fields that exist in both files
- The following 3 `hebrew` fields appear in `markers.csv`: `non_urban_intersection_hebrew`, `street1_hebrew`, `street2_hebrew` 
- Fields that exist only in `markers` and not in `hebrew`: `title`, `created`, `description`, `address`, `mainStreet`, `secondaryStreet`, `junction`.  How come they are not in `markers_hebrew` Per Atalya, the reason is historic, and these fields are not interesting enough.  **TODO** - to confirm.
- `accident_timestamp` appears only in `markers_hebrew` and not `markers`.  Since continuing working on `markers`, it's OK.

Main info() values of `markers` without hebrew:

    id                                           720541 non-null int64
    provider_and_id                              720541 non-null int64
    provider_code                                720541 non-null int64
    file_type_police                             55400 non-null float64
    accident_type                                720541 non-null int64
    accident_severity                            720541 non-null int64
    location_accuracy                            720541 non-null int64
    road_type                                    720541 non-null int64
    road_shape                                   720541 non-null int64
    day_type                                     720541 non-null int64
    police_unit                                  720541 non-null int64
    one_lane                                     720541 non-null int64
    multi_lane                                   720541 non-null int64
    speed_limit                                  720541 non-null int64
    road_intactness                              720541 non-null int64
    road_width                                   720541 non-null int64
    road_sign                                    720541 non-null int64
    road_light                                   720541 non-null int64
    road_control                                 80906 non-null float64
    weather                                      720541 non-null int64
    road_surface                                 720541 non-null int64
    road_object                                  720541 non-null int64
    object_distance                              720541 non-null int64
    didnt_cross                                  720541 non-null int64
    cross_mode                                   720541 non-null int64
    cross_location                               720541 non-null int64
    cross_direction                              720541 non-null int64
    road1                                        154049 non-null float64
    road2                                        146383 non-null float64
    km                                           140626 non-null float64
    km_raw                                       140626 non-null float64
    km_accurate                                  140626 non-null object
    yishuv_symbol                                720541 non-null int64
    yishuv_name                                  560665 non-null object
    geo_area                                     720541 non-null int64
    day_night                                    720541 non-null int64
    day_in_week                                  720541 non-null int64
    traffic_light                                720541 non-null int64
    region                                       605597 non-null float64
    district                                     720541 non-null int64
    natural_area                                 605597 non-null float64
    municipal_status                             604357 non-null float64
    yishuv_shape                                 605597 non-null float64
    street1                                      553240 non-null float64
    street1_hebrew                               389936 non-null object
    street2                                      71838 non-null float64
    street2_hebrew                               70713 non-null object
    house_number                                 487549 non-null float64
    non_urban_intersection                       34382 non-null float64
    non_urban_intersection_hebrew                33018 non-null object
    non_urban_intersection_by_junction_number    33136 non-null object
    urban_intersection                           70730 non-null float64
    accident_year                                720541 non-null int64
    accident_month                               720541 non-null int64
    accident_day                                 720541 non-null int64
    accident_hour_raw                            720541 non-null int64
    accident_hour                                720541 non-null int64
    accident_minute                              720541 non-null int64
    geom                                         710613 non-null object
    longitude                                    710613 non-null float64
    latitude                                     710613 non-null float64
    x                                            710613 non-null float64
    y                                            710613 non-null float64

`location_accuracy` field:

In [19]:
m_all[m_all['x'].isnull()]['location_accuracy'].value_counts()

9    9928
Name: location_accuracy, dtype: int64

**Conclusion**: every time *location_accuracy* is 9, there will be no info for *geom*, *longitude*, *latitude*, *x*, *y*

`file_type_police` field:

In [24]:
m_pol['file_type_police'].value_counts(dropna=False)

NaN    142024
1.0     12245
3.0       312
Name: file_type_police, dtype: int64

In [25]:
m_no_pol['file_type_police'].value_counts(dropna=False)

NaN    523117
3.0     42843
Name: file_type_police, dtype: int64

`file_type_police` doesn't appear in `Fields_Meaning_for_2008-2018` doc.

It's new from last year or two, this is the decision about file type that was initially provided by the police, real value is given by the Lamas in `provider_code`, sometimes the decision is different from `provider_code`.  
**TODO** - put in Excel of fields explanation.
**TODO** - check the connection between the 2 fields.

`road_control` field:

In [26]:
m_all[m_all['road_type']==1]['road_control'].value_counts(dropna=False)

NaN    26606
2.0    20226
1.0     7759
7.0     7613
6.0     6684
5.0     4821
3.0      412
4.0      112
0.0        2
Name: road_control, dtype: int64

In [27]:
m_all[m_all['road_type']==3]['road_control'].value_counts(dropna=False)

NaN    20711
2.0     9061
7.0     1952
1.0     1645
5.0      966
6.0      902
3.0       89
4.0       52
Name: road_control, dtype: int64

**Question:** It says that `road_control` appears only for intersections.  Practically, even for intersections, it doesn't appear most of the time.

**TODO**: check if maybe until a certain year it didn't appear

`road1` field:

In [40]:
print(f"There are {m_all[m_all['road_type']==3].shape[0]} road_type 3 (intersection out of city) observations, out of which road1 is {m_all[m_all['road_type']==3]['road1'].isnull().sum()} Null")

There are 35378 road_type 3 (intersection out of city) observations, out of which road1 is 47 Null


In [33]:
m_all[m_all['road_type']==3]['road1'].value_counts().head()

4.0     6247
40.0    2256
70.0    2222
65.0    1731
20.0    1543
Name: road1, dtype: int64

In [43]:
print(f"There are {m_all[m_all['road_type']==4].shape[0]} road_type 4 (not intersection out of city) observations, out of which road1 is {m_all[m_all['road_type']==4]['road1'].isnull().sum()} Null")

There are 123379 road_type 4 (not intersection out of city) observations, out of which road1 is 4661 Null


**QUESTION** / **TODO**: can we use this field and either impute these values or drop them? **TODO:** check location_accuracy

In [37]:
print(f"There are {m_all[m_all['road_type']==3].shape[0]} road_type 3 (intersection out of city) observations, out of which road2 is {m_all[m_all['road_type']==3]['road2'].isnull().sum()} Null")

There are 35378 road_type 3 (intersection out of city) observations, out of which road2 is 456 Null


`km` field:

In [44]:
print(f"There are {m_all[m_all['road_type']==3].shape[0]} road_type 3 (intersection out of city) observations, out of which km is {m_all[m_all['road_type']==3]['km'].isnull().sum()} Null")

There are 35378 road_type 3 (intersection out of city) observations, out of which km is 1017 Null


In [45]:
print(f"There are {m_all[m_all['road_type']==3].shape[0]} road_type 3 (intersection out of city) observations, out of which km_raw is {m_all[m_all['road_type']==3]['km_raw'].isnull().sum()} Null")

There are 35378 road_type 3 (intersection out of city) observations, out of which km_raw is 1017 Null


In [48]:
(m_all[m_all['road_type']==3]['km'] - m_all[m_all['road_type']==3]['km_raw']).value_counts(dropna=False)

0.0    34361
NaN     1017
dtype: int64

In [50]:
(m_all[m_all['road_type']==3]['km'] - m_all[m_all['road_type']==3]['km_accurate']).value_counts(dropna=False)

TypeError: unsupported operand type(s) for -: 'float' and 'str'

**Question**: `km_raw` seems to be exactly like km, why do we have it. **Answer:** use km real, '-' in the beginning, accurate says if has '-'